### PR TITLE
KAFKA-17153 KafkaMetricsGroup#newGauge should accept functional interface instead of `com.yammer.metrics.core.Gague`

### PR DESCRIPTION
**Pull request by brandboat**
related to https://issues.apache.org/jira/browse/KAFKA-17153,

Java disallow lambdas to implement abstract classes, and hence there are many ugly impl of Gague in code base. Those bad looking code will spring up as we migrate more and more scala code to java.

Hence, we should offer a variety to accept functional interface for java code. For example:

```java
    public final <T> Gauge<T> newGauge(String name, Supplier<T> metric) {
        return newGauge(name, new Gauge<T>() {
            @Override
            public T value() {
                return metric.get();
            }
        }, Collections.emptyMap());
    }
```

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by chia7712** [comment]
@brandboat those overloaded methods are final, and mockito may call the real method if mockito can't mock them. In order to make those tests pass, we can check the method which is mocked correctly. That is a known limit, so it is fine to change the test if we don't want to remove the "final" declaration.

**Comment by chia7712** [comment]
> if we don't want to remove the "final" declaration.

or we can remove the `final`. not sure why this method needs those final methods when we try to mock it.

**Comment by brandboat** [comment]
> or we can remove the final. not sure why this method needs those final methods when we try to mock it.

Thanks for the clear explanation, I will remove the `final` in newGauge methods.

File: server-common/src/main/java/org/apache/kafka/server/metrics/KafkaMetricsGroup.java [hunk 1]
```diff
@@ -70,16 +71,26 @@ public static MetricName explicitMetricName(String group, String typeName,
         return new MetricName(group, typeName, name, scope, nameBuilder.toString());
     }
 
-    public final <T> Gauge<T> newGauge(String name, Gauge<T> metric, Map<String, String> tags) {
-        return KafkaYammerMetrics.defaultRegistry().newGauge(metricName(name, tags), metric);
+    public <T> Gauge<T> newGauge(String name, Supplier<T> metric, Map<String, String> tags) {
+        return KafkaYammerMetrics.defaultRegistry().newGauge(metricName(name, tags), new Gauge<T>() {
```
Comment (chia7712) [inline server-common/src/main/java/org/apache/kafka/server/metrics/KafkaMetricsGroup.java]: how about `return newGauge(metricName(name, tags), metric);`?

File: storage/src/main/java/org/apache/kafka/storage/internals/log/RemoteStorageThreadPool.java [hunk 2]
```diff
@@ -51,18 +50,10 @@ public String logPrefix() {
...
-            }
-        });
-        metricsGroup.newGauge(REMOTE_LOG_READER_AVG_IDLE_PERCENT_METRIC.getName(), new Gauge<Double>() {
-            @Override
-            public Double value() {
-                return 1 - (double) RemoteStorageThreadPool.this.getActiveCount() / (double) RemoteStorageThreadPool.this.getCorePoolSize();
-            }
-        });
+        metricsGroup.newGauge(REMOTE_LOG_READER_TASK_QUEUE_SIZE_METRIC.getName(),
+                () -> RemoteStorageThreadPool.this.getQueue().size());
```
Comment (chia7712) [inline storage/src/main/java/org/apache/kafka/storage/internals/log/RemoteStorageThreadPool.java]: how about `getQueue().size()` ?

File: storage/src/main/java/org/apache/kafka/storage/internals/log/RemoteStorageThreadPool.java [hunk 3]
```diff
@@ -51,18 +50,10 @@ public String logPrefix() {
...
+        metricsGroup.newGauge(REMOTE_LOG_READER_AVG_IDLE_PERCENT_METRIC.getName(),
+                () -> 1 - (double) RemoteStorageThreadPool.this.getActiveCount() / (double) RemoteStorageThreadPool.this.getCorePoolSize());
```
Comment (chia7712) [inline storage/src/main/java/org/apache/kafka/storage/internals/log/RemoteStorageThreadPool.java]: ditto

**Review by chia7712 — COMMENTED** [review]
@brandboat thanks for this nice refactor!

File: server-common/src/main/java/org/apache/kafka/server/metrics/KafkaMetricsGroup.java (hunk 1 above)
Comment (brandboat) [reply server-common/src/main/java/org/apache/kafka/server/metrics/KafkaMetricsGroup.java]: Oops.

**Review by chia7712 — APPROVED** [review]
LGTM

**Pull request merged**